ros/src/kinematic_icp_ros/server/LidarOdometryServer.cpp

```diff
@@ -184,18 +184,26 @@ void LidarOdometryServer::RegisterFrame(const sensor_msgs::msg::PointCloud2::Con
     // Extract timestamps
     const auto timestamps = GetTimestamps(msg);
     const auto &[min_it, max_it] = std::minmax_element(timestamps.cbegin(), timestamps.cend());
-    // From double to ROS TimeStamp
+
+    // Todo: Move everything to a separate unit
     auto toStamp = [](const double &time) -> builtin_interfaces::msg::Time {
         return rclcpp::Time(tf2::durationFromSec(time).count());
     };
-    // Update what is the current stamp of this iteration
-    const auto begin_scan_stamp = min_it != timestamps.cend() ? toStamp(*min_it) : last_stamp;
-    const auto end_scan_stamp = max_it != timestamps.cend() ? toStamp(*max_it) : msg->header.stamp;
-    current_stamp_ = end_scan_stamp;
+    auto toTime = [](const builtin_interfaces::msg::Time &stamp) -> double {
+        return rclcpp::Time(stamp).nanoseconds() * 1e-9;
+    };
+
+    // Get scan duration and current stamp
+    const auto begin_scan_time = min_it != timestamps.cend() ? *min_it : toTime(last_stamp);
+    const auto end_scan_time = max_it != timestamps.cend() ? *max_it : toTime(msg->header.stamp);
+    const double scan_duration = std::abs(end_scan_time - begin_scan_time);
+    current_stamp_ = begin_scan_time < toTime(last_stamp)
+                         ? toStamp(toTime(last_stamp) + scan_duration)
```

Comment:
Here, we assume that the current scan begins as soon as the last one ends. Of course, this does not hold in case of frame drops or if the scanner does not measure any points at the beginning or end of the scanning duration, similar to the problem discussed in https://github.com/PRBonn/kiss-icp/issues/209 for deskewing.

Alternatively, we could replace `toTime(last_stamp)` with `toTime(msg->header.stamp)` and assume that if relative timestamps are given, the message header must be timestamped at the beginning of the scan.

Or we assume that the relative timestamps would be negative if message header timestamping happens at the end of the scan. In this case, we could get the `current_stamp_` by `toStamp(toTime(msg->header.stamp) + scan_time_difference)`.